src/Particle/Par_MapMesh2Particles.cpp

```diff
@@ -0,0 +1,261 @@
+#include "GAMER.h"
+
+#ifdef PARTICLE
+
+void Par_MapMesh2Particles ( const int lv, const int P, const double EdgeL[3],
+                             const double EdgeR[3], const int AttrSize3D, const real *Attr,
+                             const int NPar, real *InterpParPos[3],
+                             const real ParType[], const long ParList[],
+                             bool useTracers, real ParAttr[], const int ParGhost,
+                             const bool CorrectVelocity )
+{
+
+   typedef real (*vla)[AttrSize3D][AttrSize3D][AttrSize3D];
+   vla Attr3D = ( vla )Attr;
+
+   const ParInterp_t IntScheme    = amr->Par->InterpTracer;
+
+   const double _dh               = 1.0/amr->dh[lv];
+
+   for (int p=0; p<NPar; p++)
+   {
+      long ParID = ParList[p];
+
+      if ( useTracers )
+      {
+//       skip massive particles
+         if ( ParType[ParID] != PTYPE_TRACER )
+            continue;
+      }
+      else
+      {
+//       skip tracer particles
+         if ( ParType[ParID] == PTYPE_TRACER )
+            continue;
+      }
+
+      switch ( IntScheme ) {
+
+//    1 NGP
+      case ( PAR_INTERP_NGP ):
+      {
+         int idx[3];
+
+//       calculate the nearest grid index
+         for (int d=0; d<3; d++)
+         {
+            idx[d] = int( ( InterpParPos[d][p] - EdgeL[d] )*_dh );
```

Comment:
I do agree--I think that this is ok, since we do have the data stored in `Attr` to do the mapping. I think we do want the attributes of the particles to reflect their current locations when we leave this routine, yes?